[OpenFX]: add version for Davinci Resolve 18.x

ramajd opened the pull request:
- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [ ] The "supports" clause reflects platforms that may be fixed by this new version.
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

ramajd committed:
[OpenFX]: add version pointing the Davinci Resolve 18.x

ramajd commented:
this pull request contains the changes required to apply on the official OpenFx 1.4 port, to match with the modified version shipped with Davinci Resolve 18.x.

LilyWangLL commented on ports/openfx/vcpkg.json:
```diff
@@ -1,7 +1,7 @@
   "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg-tool/main/docs/vcpkg.schema.json",
   "name": "openfx",
-  "version": "1.4",
+  "version": "1.4-davinci-18.x",
```
```suggestion
  "version": "1.4",
  "port-version": 1,
```

LilyWangLL replied:
If you only want to download the file `ofxDrawSuite.h`, you just need to modify `port-version`.

ramajd replied:
@LilyWangLL  the purpose that I added a new tag for this change is because I wanted to keep both versions available, the original version and the modified one which is shipped with Davinci Resolve suite.

LilyWangLL commented on ports/openfx/portfile.cmake:
```diff
@@ -1,4 +1,5 @@
 string(REPLACE "." "_" UNDERSCORE_VERSION "${VERSION}")
+string(REPLACE "-davinci-18_x" "" UNDERSCORE_VERSION "${UNDERSCORE_VERSION}")
```
```suggestion
```

LilyWangLL replied:
Since here are substituting `-davinci-18_x` with a null value, why do we still need to add `-davinci-18_x` in the version?

ramajd replied:
since we are preparing the target tag to build the port from, based on the version in `vcpkg.json` file. I had to remove the `-davinci-18_x` postfix from the version string in davinci-based versions.

LilyWangLL commented:
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

dg0yt commented:
Is it really enough to add a singler header, from a much later revision?

FTR the `1.4-xyz` version pattern is not suitable because for sorting it will be "older" than `1.4`.

ramajd commented:
> Is it really enough to add a singler header, from a much later revision?
> 

@dg0yt yes, comparing the official OpenFx tag with the modified version this was the change applied on the later one. 


> FTR the `1.4-xyz` version pattern is not suitable because for sorting it will be "older" than `1.4`.

so what is the preferred naming format for the modified version and keep both versions available?

dg0yt commented:
> so what is the preferred naming format for the modified version and keep both versions available?

There is no perfect solution in scheme `version`, from my observations in other ports. The suggestion to only increase `port-version` is the best approach to start with.

But this still creates a different package than the official `1.4`. I wonder if the installation of the extra header might be controlled by an extra feature `drawsuite`, with a clear description. So users can request it when they need it. But they get the official 1.4 installation by default.

OTOH the header (2023) is much newer than 1.4 (2015), and there aren't any new releases... I already commented on the age when the port was added recently. An alternative to adding a single header is to pick the whole repo at a a newer revision, and switch to scheme `version-date`.

Disclaimer: I'm only a community member.

ramajd committed:
[OpenFX]: add davinci modifications as a feature

ramajd commented:
as @dg0yt suggested I added the modifications related to the Davinci Resolve, as a feature to the official port definition, and reverted the custon `-davinci-18_x` postfix from the version string.

BillyONeal commented:
This sounds like an alternative rather than a feature. What happens to folks who were expecting the unmodified header and accidentally get DaVinci Resolve 18's custom one?

>so what is the preferred naming format for the modified version and keep both versions available?

In general, there isn't one.

See also: https://github.com/AcademySoftwareFoundation/openfx/issues/143

dg0yt commented:
> What happens to folks who were expecting the unmodified header and accidentally get DaVinci Resolve 18's custom one?

The package has no unmodified header.

ramajd commented:
> What happens to folks who were expecting the unmodified header and accidentally get DaVinci Resolve 18's custom one

@BillyONeal the davinci's feature is disabled by default. so anyone needs the default doesn't face the davinci version accidentally. 

the point it that in the modifed Davinci version, there is a header file from version 1.5 shipped with the official 1.4 OpenFx + some path modifications which won't cause any issue to the development  structure.

ramajd commented:
another approach would be to fork from the original `OpenFx` repository, apply the _Davinci Resolve_ changes related to each version, and release that as a separated package name like `davinci-openfx-sdk`. 

@BillyONeal is this an acceptable approach to follow?

BillyONeal commented:
> the point it that in the modifed Davinci version, there is a header file from version 1.5 shipped with the official 1.4 OpenFx + some path modifications which won't cause any issue to the development structure.

Can you explain why this makes sense to be a feature and/or point to how upstream thinks of this in their versioning scheme and/or what makes Davinci special as a consumer here?

JavierMatosD commented:
@ramajd, I believe the path forward here is to get these changes incorporated into upstream. Once they are incorporated we can switch this port to `version-date` and update the port. (Unless they mint and publish and new release for us. In which case we would just update the port to the new version.

ramajd commented:
@BillyONeal  @JavierMatosD The point is that [BlackMagic design](https://www.blackmagicdesign.com/) applies a series of modifications to the original SDK, specific to their own host application (Davinci Resolve). these modifications are in form of  one of the following formats:

- additional files that allow the plugin to work with host application specific features (these files end with `...Ext.h`)
- Using a combination of features from a recent OpenFX version, alongside with the old OpenFX 1.4 (like `ofxDrawSuite.h`)
- modifications to the original OpenFX 1.4 flags and parameters which is specific to the Davinci Resolve

so in order to have access both to the upstream OpenFX and the Davinci-modifiet versions, we initially proposed to add the modifications with a feature added to the upstream package.

the other suggestion we are thinking of, was to release the davinci-modified version with a totally different name (davinci-openfx-sdk), and take care of it's versioning by ourselves.

garyo commented:
Hi folks; I'm chair of the OpenFX steering committee. I just saw this, and say it's great you are doing this! I just want to let you know we're closing in on a 1.5 release of OpenFX, which should be available by late July. Our main branch is currently quite close to release status, with the exception of a couple of upcoming features. We have integrated the Resolve GPU extensions into the core spec, as well as lots of other good stuff. We also now have a cmake+conan build setup with CI (no vcpkg version -- if you'd like to contribute one, we'd be open to that!). 

If you're making an OpenFX distribution for vcpkg, I suggest using our main branch and tracking it until at least 1.5. If you want to include non-standard extensions (such as Resolve's), according to the spec all such extensions should be in a separate header file and should be properly namespaced to prevent collisions; see https://openfx.readthedocs.io/en/main/Reference/ofxStructure.html#extending-ofx

Please reach out on our Slack or Discord if you have any questions or comments!
